venvctrl/venv/relocate.py

```diff
@@ -29,7 +29,9 @@ def relocate(self, destination):
 
         for binfile in self.bin.files:
 
-            if binfile.shebang and 'python' in binfile.shebang:
+            if binfile.shebang and (
+                    'python' in binfile.shebang or 'pypy' in binfile.shebang
+            ):
 
                 binfile.shebang = '#!{0}'.format(
                     os.path.join(destination, 'bin', 'python')
```

Comment:
Keeping this line as-is results in shebangs being rewritten from `pypy` to `python`. Was that the intent or should these continue to use the `pypy` reference?

Reply:
That is the intent because `python` should refer to the local python compiler (can be different versions of python or, in my case, pypy).

Reply:
What are the conditions that lead to a file in the virtualenv with `pypy` in the shebang instead of `python`. As I understand, the files are wrapped/annotated by pip such that they always reference `python`. It sounds like something else is breaking to cause this condition.

If a shebang had `pypy` in the path then, I think, it would invalidate the virtualenv by not making references to variables managed by the environment. For example, all other files would work as expected but the one with `pypy` in it would call out to the system `pypy` instead of using the virtualenv.

I'm not suggesting this change isn't valid. I'm only curious how this problem arises.

Reply:
In our environment, we are using the PyPy executable from the portable version. When built, the RPM references such PyPy version since we provide it through the `.json` schema. 
When we install the built RPM, the shebangs that use PyPy are not updated at all. In fact, they still refer to the buildroot.

Reply:
Hrm, I'm still curious how that is not a bug in either PyPy or virtualenv if it allows for the shebangs within a venv to be created referencing anything other than `python`.

In any case, this toolchain should work in an expected way for any valid Python runtime. Practically speaking there's really only cPython and PyPy so I'm not against having `pypy` embedded in the code like this. Should another PR/Issue come across for yet another variant of Python then we can adjust some of the call signatures to make these values configurable. For now, I _think_ it's safe to just rewrite all `pypy` references in the venv to `python` since the venv is configured to route calls to `python` to the correct interpreter.

I would ask for you to add a test for this case. There's already a test looking for this behaviour here: https://github.com/kevinconway/venvctrl/blob/master/tests/test_virtual_environment.py#L53. I might recommend using the `abspath` property of the venv to inject a custom binfile for the test that has a `pypy` based shebang. The remaining portion of the test should automatically cover your use case if you do that.

Reply:
Gotcha. I updated the test as you suggested. Though, I think that a better way to do it would be to add another build job with `PyPy` as interpreter.  :slightly_smiling_face:

Reply:
@Fumaloko92 The problem, though, is that running tests under PyPy doesn't recreate this issue: https://travis-ci.org/kevinconway/venvctrl/builds/399415413. That's part of why I've been asking questions. A well behaved virtualenv creation does not result in a `pypy` shebang. The virtualenv masks that with a generic `python` name.

I believe something else in the build process is broken that results in these strange files. For example, if you were to run those `pypy` files with the virtualenv activated then they'd effectively escape the venv and use a different interpreter than intended. Whether packaged in an RPM or otherwise, I believe any shebang for Python code that does not have `python` in the path breaks the isolation of the venv.

I'm open to learning more about how this comes about in the wild and am still not opposed to the change. I should get to the merging, tagging, and releasing tomorrow.

Reply:
Could it be possible that `rpmvenv` does not handle `PyPy` properly and therefore does not set the shebangs to `python`?